### PR TITLE
Rename x11 platform to 'linuxbsd'

### PR DESCRIPTION
**Pull request by ddevault**
Also renames OS_X11 to OS_LinuxBSD, and a few other similar changes.

### CODE REVIEW

**Comment by ddevault** [comment]
Sorry about the noise while I sorted out Travis issues. This should be ready for review.

**Comment by shartte** [comment]
Why is the new platform not called POSIX? Isn't that what it is programmed against? Unix is after all trademarked :-)

File: modules/mono/config.py
Comment (shartte) [inline modules/mono/config.py]: Is this an oversight? Shouldn't this also be generic_unix?

File: modules/thekla_unwrap/SCsub
Comment (shartte) [inline modules/thekla_unwrap/SCsub]: Is this an oversight? Shouldn't this also be generic_unix?

Comment (ddevault) [reply modules/thekla_unwrap/SCsub]: Aye, another oversight. Thanks!

**Comment by ddevault** [comment]
>Why is the new platform not called POSIX? Isn't that what it is programmed against?

Well, POSIX doesn't specify a display server like x11 or Wayland. Other code here, like power stuff, crash handler, etc, aren't specified by POSIX.

`OS_Unix` (which already exists) is a bit closer to `OS_POSIX`, but a lot of platforms depend on it so it seems wiser to take the approach of least risk of introducing a "generic unix" flavor which is closer to the behavior of Unix than the more specialized derivatives like macOS and Android.

**Comment by ddevault** [comment]
>I.e. you renamed crash_handler_x11.cpp to crash_handler_unix.cpp, but that file is using a Linux specific approach to getting the stacktrace.

It's actually worse - this is a glibc-specific way of getting the stack trace. See this ticket:

https://github.com/godotengine/godot/issues/20037

I imagine that in the long run we should use feature detection at build-time to enable or disable specific implementations of this sort of thing. Or we could use runtime guesswork like we do in PowerUnix - which could easily be adapted for e.g. BSD in the future. I think that needn't be solved in this particular patch, though.

**Comment by shartte** [comment]
Yeah the crash handler aspect has always been tough to figure out. I'd have to check out a recent version of Google's breakpad to look at what they're doing now. Maybe they have a better abstraction?

**Comment by ddevault** [comment]
imo the crash handler is nice-to-have at best - loading up a core dump is going to give you more useful information and tools to work with anyway. I used to have crash handlers in my programs but eventually removed them. Do you think solving this problem is a blocker for this patch?

**Comment by shartte** [comment]
Oh I am in no position to actually decide any of that heh. I was just wondering whether you had a concept for handling sub-platform specific code in generic_unix, such as this crash handler.

**Comment by ddevault** [comment]
As the need for that grows, we can start writing some `platform/generic_unix/`-specific interfaces to abstract those things and add implementations in e.g. `platform/generic_unix/linux` or even `platform/generic_unix/glibc/`.

**Comment by aaronfranke** [comment]
I much prefer the name `linuxbsd` which explicitly describes what it actually is. "generic_unix" makes me think of Unix from the 1970s. Linux and BSD are not Unix, they are Unix-like. 

Neither Linux nor BSD are certified Unix, but [Mac and Solaris are](https://www.opengroup.org/openbrand/register/). It is very misleading to make the name containing "unix" when it doesn't actually run on any certified Unix systems.

**Comment by ddevault** [comment]
Calling it linuxbsd would be misleading as well, and doesn't describe it for what it is either. I don't see any reason why any of the code here wouldn't run on Minix, OpenSolaris, macOS, or any of several more niche systems. No one takes Unix certification seriously, and none of the code here depends on the niche Unix features which are unique to certified Unicies. "Generic Unix" is used all over the internet to refer to generic Unix-like systems, for which Linux and BSD certainly qualify.

That being said, we don't need to bikeshed this to death. If you insist on a particular name I'll update the patch to use it.

**Comment by aaronfranke** [comment]
The final decision is not mine, it will probably be @akien-mga 's.

**Comment by Megalomaniak** [comment]
I'll just add here that I've seen users on forums be confused about what build to download for linux before because of the x11 naming. Not sure how much better calling it generic unix is going to be but I expect it to be at least an improvement.

**Comment by toger5** [comment]
I checked the code real quick and it looks good. Didn't check every singe function tough, but the general idea of separating the DisplayDriver implementations by using inheritance is good Imo. 
It would be amazing if @hpvb could also give an opinion on the whole topic.

**Comment by vnen** [comment]
> Calling it linuxbsd would be misleading as well, and doesn't describe it for what it is either. I don't see any reason why any of the code here wouldn't run on Minix, OpenSolaris, macOS, Cygwin, or any of several more niche systems.

I don't think it is misleading because it is meant to run on Linux and *BSD. While it _could_ run on other systems, it's not meant to, so you expect to experience incompatibilities.

In any case, I think that opening a PR (worse: many PRs) just to decide a naming convention is a waste of time. We should discuss on an issue until we reach a conclusion and then change the code for good. No use making the same changes for each name suggestion, just to later discard all the rejected ones.

**Comment by ddevault** [comment]
>I don't think it is misleading because it is meant to run on Linux and *BSD. While it could run on other systems, it's not meant to, so you expect to experience incompatibilities.

What part of this is only meant to run on Linux or *BSD? Only some of the code is Linux-specific and is optional. The code will compile and run on any POSIX system with an X11 server.

>In any case, I think that opening a PR (worse: many PRs) just to decide a naming convention is a waste of time

That's not what this PR is for. This PR also includes a considerable amount of cleanup from the previous WIP PR, which was not in a mergable state.

**Comment by santouits** [comment]
Compiled with both llvm and gnu and didn't find any regression in my game. 

About the name, does posix_freedesktop describe the abstractions that were made?

**Comment by ddevault** [comment]
>About the name, does posix_freedesktop describe the abstractions that were made?

No. POSIX would be more appropriate for `drivers/unix/`. POSIX doesn't describe a display server of any kind. Freedesktop doesn't either, Freedesktop.org is just the name of a website which provides tools which are used by XDG, a standards body which mostly decides standards for window managers, and hosts some of the code for Xorg server et al - like a GitHub for desktop developers. The group which maintains the predominant X11 implementation for Linux is X.org, which is separate still from Freedesktop, but still not right - Xorg != X11, a protocol which is governed separately and for which there are many implementations.

I'd sooner have linuxbsd than freedesktop, but both are wrong.

**Comment by bruvzg** [comment]
> What part of this is only meant to run on Linux or *BSD? Only some of the code is Linux-specific and is optional. The code will compile and run on any POSIX system with an X11 server.

You can build and run X11 version of Godot on macOS (with exception of webm/webp modules, but these two fail on any "non-standard" configuration) by setting pkg-config search path to XQuartz location, and applying following minimal patch.

<details>
<summary>Patch</summary>

```patch
diff --git a/platform/x11/detect.py b/platform/x11/detect.py
index b5ad59e60..1439adddb 100644
--- a/platform/x11/detect.py
+++ b/platform/x11/detect.py
@@ -14,7 +14,7 @@ def get_name():
 
 def can_build():
 
-    if (os.name != "posix" or sys.platform == "darwin"):
+    if (os.name != "posix"):
         return False
 
     # Check the minimal dependencies
diff --git a/platform/x11/platform_config.h b/platform/x11/platform_config.h
index 9e6011ddf..21a2a6b6a 100644
--- a/platform/x11/platform_config.h
+++ b/platform/x11/platform_config.h
@@ -28,9 +28,12 @@
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
 /*************************************************************************/
 
-#ifdef __linux__
+#if defined (__linux__) || defined(__APPLE__)
 #include <alloca.h>
 #endif
+#if defined(__APPLE__)
+#define PTHREAD_RENAME_SELF
+#endif
 #if defined(__FreeBSD__) || defined(__OpenBSD__)
 #include <stdlib.h>
 #define PTHREAD_BSD_SET_NAME
```

</details>

<details>
<summary>Screenshot</summary>

Screenshot 2019-03-21 at 23 45 32

</details>

**Comment by ddevault** [comment]
I think the case in favor of each proposed name has been made. Can someone in a position to make decisions choose the name they prefer, and I'll update this patch accordingly? Will also rebase #27463.

**Comment by reduz** [comment]
GenerixUnix is a terrible name, I think the best would be Freedesktop which is actually the real platform.

That said, @hpvb is planning a massive restructuration of this, so not sure what will be done with this.

**Comment by ddevault** [comment]
As I explained [here](https://github.com/godotengine/godot/pull/27266#issuecomment-475365618), freedesktop is *not* the name of the platform. But again, I just want a decision. If you're deciding that the port shall be called freedesktop, then freedesktop it shall be.

@hpvb how can I help?

**Comment by akien-mga** [comment]
IMO, "LinuxBSD" (`platform/linuxbsd`) is the best option. Sorry to be blunt, but we don't really care about other Unix-likes which are even more niche than Linux and BSD (and I say that as a daily Linux user and gamer, and strong advocate of Linux as a gaming platform). Godot is a game engine, so our scope is platforms where people actually develop and play games.

"LinuxBSD" is clear for end users and contributors that this is the platform code for Linux and *BSD. It can of course be made to support other Unixes, and those will have to do with being labelled wrongly, but that's not a big deal IMO. We already have a `platform:linuxbsd` GitHub label for `platform/x11`, so IMO it makes the most sense. (It used to be `platform:linux`, but we appended `bsd` a while ago.)

I'm fine with using "Freedesktop" if we really want to have an accurate-ish common denominator for all the platforms supported by this port, since the code in this port is mainly limited to Freedesktop (and XDG) related code. The Unix specific code lives in another place (`drivers/unix`) and is shared by other Unix-derived ports. As such I do agree that "generic_unix" is not a good name for this platform.

AFAICT "Freedesktop" and "LinuxBSD" are the only two names that core developers are somewhat OK with, so I'll just make a mini poll in the next comments. Add a :+1: to the name you prefer.

**Note:** Whatever name we settle on, keep it mind that it will only be relevant in the port's code and the name of the `platform/` port folder:
- For the export templates, we will export end user friendly names and likely provide different templates for the different platforms supported from this code (so a "Linux" preset, a "FreeBSD" preset, an "OpenBSD" preset, etc.). Even if the platform code is almost the same apart from a couple platform-specific defines, the export templates binaries can't be, so to support more platforms than just Linux we'll have to provide new binaries and presets.
- For the SCons platform option, we can add as many aliases as we want if it feels weird to type `scons p=linuxbsd` or `scons p=freedesktop`. Aliases like `linux`, `bsd` or `fdo` are trivial to add.

**Comment by akien-mga** [comment]
If you favor the "LinuxBSD" (`platform/linuxbsd`) name, :+1: this comment.

**Comment by akien-mga** [comment]
If you favor the "Freedesktop" (`platform/freedesktop`) name, :+1: this comment.

**Comment by wmww** [comment]
I've consulted @ddevault and we've got a plan. He'll refactor this PR to use `LinuxBSD` and remove the bit of it that splits up OS and display code. It will then simply be a platform rename. I'll rebase #27747 on top of it, which should improve and clean my PR up.

**Comment by ddevault** [comment]
>I've consulted @ddevault and we've got a plan. He'll refactor this PR to use LinuxBSD and remove the bit of it that splits up OS and display code. It will then simply be a platform rename. I'll rebase #27747 on top of it, which should improve and clean my PR up.

Done.

**Comment by wmww** [comment]
Nice work. This now simply improves the platform name with limited risk and no architectural changes. I hope it can move forward so we can start looking at #27747, which builds on top of it.

File: platform/linuxbsd/export/export.cpp
Comment (hpvb) [inline platform/linuxbsd/export/export.cpp]: This should not be changed in this PR. The user-facing name should probably remain Linux for now.

File: platform/linuxbsd/power_linuxbsd.cpp
Comment (hpvb) [inline platform/linuxbsd/power_linuxbsd.cpp]: I'm not sure this code even works on BSDs. FreeBSD only has /proc for Linux compatibility iirc?

**Review by hpvb — CHANGES_REQUESTED** [review]

**Comment by Calinou** [comment]
It should be noted in the changelog that this is a compatibility-breaking change, as the feature flag will no longer be `X11` and the OS name will no longer be `Linux/X11` (as returned by `OS.get_name()`).

File: platform/linuxbsd/power_linuxbsd.cpp
Comment (wmww) [reply platform/linuxbsd/power_linuxbsd.cpp]: Weather it works or not, it *is* used on all X11 platforms (as the old name suggests). Therefor, I think this is an appropriate name.

Comment (ddevault) [reply platform/linuxbsd/power_linuxbsd.cpp]: It:

1. Works on non-X11 displays, like Wayland
2. Doesn't *break* on BSD
3. Is already somewhat generic and adding BSD support would be easy

So I agree with @wmww that the name here is right.

**Comment by ddevault** [comment]
>It should be noted in the changelog that this is a compatibility-breaking change, as the feature tag will no longer be X11 and the OS name will no longer be Linux/X11 (as returned by OS.get_name()).

Changelog updated, though the OS name change was rolled back per @hpvb's suggestion.

**Comment by Calinou** [comment]
I didn't mean you should add it to `CHANGELOG.md` (it would be for 3.2 anyway), it was mostly a note-to-self as I'm the one maintaining the changelog :slightly_smiling_face:

File: platform/linuxbsd/export/export.cpp
Comment (aaronfranke) [reply platform/linuxbsd/export/export.cpp]: If it is changed, this could probably be "Linux/BSD" or "Linux&BSD", since I don't think special characters are a problem here.

**Comment by ddevault** [comment]
Rebased against the latest master and addressed merge conflicts.

**Comment by aaronfranke** [comment]
@ddevault You will need to rebase this very often, as this code changes often. The milestone has also been changed to 4.0 so it won't be merged until after 3.2. Still, I suggest rebasing often to avoid bigger monolithic rebases later and to continually test this.

**Comment by ddevault** [comment]
I kept up with it for a while but without any feedback or estimate of when it'll be reviewed, it's not worth my time to keep it up-to-date. I'd be happy to rebase it once it starts receiving feedback.

File: platform/linuxbsd/export/export.cpp
Comment (Anutrix) [inline platform/linuxbsd/export/export.cpp]: Just confirming. OS name is gonna be "Unix" and not "Linux/BSD"?

Comment (wmww) [reply platform/linuxbsd/export/export.cpp]: No. `linuxbsd` on account of there already being a `unix` OS (that includes MacOS)

Comment (Anutrix) [reply platform/linuxbsd/export/export.cpp]: So line 54 will be changed?

Comment (ddevault) [reply platform/linuxbsd/export/export.cpp]: Yes.

File: SConstruct
Comment (aaronfranke) [inline SConstruct]: Should also alias "x11" to "linuxbsd" so that exiting build scripts continue to work.

**Review by aaronfranke — CHANGES_REQUESTED** [review]

**Comment by akien-mga** [comment]
Superseded by #37317, but thanks for the contribution nevertheless!

**Pull request closed without merging**
